_overviews/hello-scala/futures.md

```diff
@@ -0,0 +1,361 @@
+---
+layout: multipage-overview
+title: Scala Futures
+description: This page provides an introduction to Futures in Scala, including Future callback methods.
+partof: hello_scala
+overview-name: Hello, Scala
+discourse: true
+num: 55
+previous-page: concurrency-signpost
+next-page: where-next
+---
+
+When you want to write parallel and concurrent applications in Scala, you *could* still use the native Java `Thread` — but the Scala [Future](https://www.scala-lang.org/api/current/scala/concurrent/Future$.html) makes parallel/concurrent programming much simpler, and it’s preferred.
+
+Here’s a description of `Future` from its Scaladoc:
+
+>“A Future represents a value which may or may not *currently* be available, but will be available at some point, or an exception if that value could not be made available.” 
+
+
+### Thinking in futures
+
+To help demonstrate this, in single-threaded programming you bind the result of a function call to a variable like this:
+
+```scala
+def aShortRunningTask(): Int = 42
+val x = aShortRunningTask
+```
+
+With code like that, the value `42` is bound to the variable `x` immediately.
+
+When you’re working with a `Future`, the assignment process looks similar:
+
+```scala
+def aLongRunningTask(): Future[Int] = ???
+val x = aLongRunningTask
+```
+
+But because `aLongRunningTask` takes an indeterminate amount of time to return, the value in `x` may or may not be *currently* available, but it will be available at some point (in the future).
+
+Another important point to know about futures is that they’re intended as a one-shot, “Handle this relatively slow computation on some other thread, and call me back with a result when you’re done” construct. (As a point of comparison, [Akka](https://akka.io) actors are intended to run for a long time and respond to many requests during their lifetime, but each future you create is intended to be run only once.)
+
+In this lesson you’ll see how to use futures, including how to run multiple futures in parallel and combine their results in a for-expression, along with other methods that are used to handle the value in a future once it returns.
+
+>Tip: If you’re just starting to work with futures and find the name `Future` to be confusing in the following examples, replace it with the name `ConcurrentTask`, which can be easier to understand initially.
```

Comment:
I disagree with using the word `Task`, which means to me something that we can repeatedly schedule or that we can decide when to run. This is not the case with `Future`, which, IMHO, is very well named, because it is just a type that will eventually hold a value, at a later point of the execution.

Reply:
I don’t have a problem deleting this, but I thought I would mention that I found the term “concurrent task” in the first paragraph of this web page:

https://docs.scala-lang.org/overviews/core/futures.html

I think the word Future is much more common now, but when I was first learning about them the phrase “concurrent task” helped me understand them.

That being said, if it’s at all confusing, let’s delete it.

Reply:
Hmm, I might be the only one finding that `Task` is misleading, so maybe it’s not that important but what would you think of `ConcurrentResult` instead?

Reply:
I’m fine removing it, or using `ConcurrentResult` instead. I don’t want anything to be misleading. Should we wait for others to review this, or just make a decision here? Thanks!

Reply:
Let’s try with `ConcurrentResult`, then! If someone has a better suggestion, I’m happy to hear it!